kendzi3d-jogl-model-render/src/main/java/kendzi/jogl/model/render/ModelRender.java

```diff
@@ -345,38 +357,65 @@ public void bindTexture(GL2 gl, Texture texture) {
         texture.bind(gl);
     }
 
-    public static void setDefaultMaterial(GL2 gl) {
+    /**
+     * @param pGl
+     *            FIXME move to util!
+     */
+    public static void enableTransparentText(GL2 pGl) {
+        // do not draw the transparent parts of the texture
+        pGl.glEnable(GL.GL_BLEND);
+        pGl.glBlendFunc(GL.GL_SRC_ALPHA, GL.GL_ONE_MINUS_SRC_ALPHA);
+        // don't show source alpha parts in the destination
+
+        // determine which areas of the polygon are to be rendered
+        pGl.glEnable(GL2ES1.GL_ALPHA_TEST);
+        pGl.glAlphaFunc(GL.GL_GREATER, 0); // only render if alpha > 0
+    }
+
+    /**
+     * @param pGl
+     *            FIXME move to util!
+     */
+    public static void disableTransparentText(GL2 pGl) {
+        pGl.glDisable(GL2ES1.GL_ALPHA_TEST);
+        pGl.glDisable(GL.GL_BLEND);
+
+    }
+
+    public void setDefaultMaterial(GL2 gl) {
+
+        int sides = drawTwoSided ? GL.GL_FRONT_AND_BACK : GL.GL_FRONT;
```

Comment:
Its look wrong, that function should reset opengl material globally, so if we use front material and back material e.g. for fence then here we need to reset both of them to some defaults we cannot leave some random values for back material.
I think always GL_FRONT_AND_BACK should be use but I do not know to good how opengl is handling front/back materials I was simply using front and never think about it :/

Reply:
Imho this is a non-issue.  If we set the material using GL.GL_FRONT, then *in any case* because we also use single sided lighting to draw the model.  Single sided means that opengl will only use the front material to draw *all* lines, polygons, faces:

This is from opengl docs:
>glMaterial assigns values to material parameters.
>>There are two matched sets of material parameters. One, the front-facing set, is used to shade points, lines, bitmaps, and all polygons (when two-sided lighting is disabled), or just front-facing polygons (when two-sided lighting is enabled).  The other set, back-facing, is used to shade back-facing polygons only when two-sided lighting is enabled.

It would hypothetically be of importance if we mix GL_FRONT and GL_FRONT_AND_BACK material setup  while leaving two sided lighting enabled.  But we never do that.

If force two sided menu option is disabled, then the EventListener will set the default material using GL_FRONT,  that material is used for *all* faces, as long as TWO_SIDED_LIGHTING is off.  Any time we turn it on, we also set the back-facing material set (isSidesChanged() will be true on such switches and GL_FRONT_AND_BACK) and only then the back-facing material set is used.

Note that you have setDefaultMaterial(GL2 gl)  and  setupDefaultMaterial(GL2 pGL)  in ModelRender.java,  only the first is called in Kendzi3dGLEventListener:
```java
        modelRender.resetMaterials();
        modelRender.setDefaultMaterial(gl);
```

Reply:
Actually, considering the opengl documentation, we should not have to track  GL_FRONT_AND_BACK -> GL_FRONT changes.  The promise of the documentation is, that in single sided lighting mode only front-facing set is used.  If the doc is to be trusted, in single sided lighting mode the back-facing set is never used  (even if it is set).

So it should suffice to re-setup material solely on the opposite flank, i.e. when moving from GL_FRONT -> GL_FRONT_AND_BACK.  Translated to the relevant code snippet in ModelRender.java this is

```java
    private boolean isSidesChanged(int sides) {
        // return lastSides == 0 || lastSides != sides;  // true on all toggle flanks
        return lastSides == 0 || (lastSides == GL.GL_FRONT && sides == GL.GL_FRONT_AND_BACK);
    }
```

Using this, it resolves light_bug1.osm as well,  but I'd rather not commit this optimization until I've received feedback from you that the obvious fix in the latest commit works..

Reply:
I've commited the optimization as well,  but please test the last two commits separately.  The behavior should be identical (except for the number of times texture setup is called), visually there should be no difference.

If you find a difference in output between last two commits  (or an issue that does not work with both)  that is related to toggling the force two sided menu option, please report back.